crates/edr_napi/src/logger.rs

```diff
@@ -0,0 +1,1235 @@
+                    gas_limit = transaction.gas_limit()
+                ),
+            );
+
+            let block_number = block_result.block.header().number;
+            logger.log_with_title(format!("Block #{block_number}"), block_result.block.hash());
+
+            logger.log_console_log_messages(&block_result.console_log_inputs);
+
+            let transaction_failure = edr_provider::TransactionFailure::from_execution_result(
+                transaction_result,
+                transaction_hash,
+            );
+
+            if let Some(transaction_failure) = transaction_failure {
+                logger.log_transaction_failure(&transaction_failure);
+            }
+        });
+    }
+
+    fn print<const REPLACE: bool>(&mut self, message: impl ToString) {
+        if !self.is_enabled {
+            return;
+        }
+
+        let formatted = self.format(message);
+
+        let status = self
+            .print_line_fn
+            .call((formatted, REPLACE), ThreadsafeFunctionCallMode::Blocking);
+
+        assert_eq!(status, napi::Status::Ok);
+    }
+
+    fn print_empty_line(&mut self) {
+        self.print::<false>("");
+    }
+
+    fn print_logs(&mut self) -> bool {
+        let logs = std::mem::take(&mut self.logs);
+        if logs.is_empty() {
+            return false;
+        }
+
+        for log in logs {
+            let line = match log {
+                LogLine::Single(message) => message,
+                LogLine::WithTitle(title, message) => {
+                    let title = format!("{title}:");
+                    format!("{title:indent$} {message}", indent = self.title_length + 1)
+                }
+            };
+
+            self.print::<false>(line);
+        }
+
+        true
+    }
+
+    fn print_method(&mut self, method: &str) {
+        if let Some(collapsed_method) = self.collapsed_method(method) {
+            collapsed_method.count += 1;
+
+            let line = format!("{method} ({count})", count = collapsed_method.count);
+            self.print::<true>(Color::Green.paint(line));
+        } else {
+            self.state = LoggingState::CollapsingMethod(CollapsedMethod {
+                count: 1,
+                method: method.to_string(),
+            });
+
+            self.print::<false>(Color::Green.paint(method));
+        }
+    }
+
+    /// Retrieves the collapsed method with the provided name, if it exists.
+    fn collapsed_method(&mut self, method: &str) -> Option<&mut CollapsedMethod> {
+        if let LoggingState::CollapsingMethod(collapsed_method) = &mut self.state {
+            if collapsed_method.method == method {
+                return Some(collapsed_method);
+            }
+        }
+
+        None
+    }
+
+    fn replace_last_log_line(&mut self, message: impl ToString) {
+        let formatted = self.format(message);
+
+        *self.logs.last_mut().expect("There must be a log line") = LogLine::Single(formatted);
```

Comment:
Yes, `replace_last_log_line` should only be called if `log` was called first. In Hardhat it would also have crashed, as `logs[logs.length - 1]` would have returned `undefined` in TS.